Fixes and docu for memory analyzer

xbauch opened the pull request.

sonodtt commented on doc/cprover-manual/memory-analyzer.md:
```diff
@@ -0,0 +1,70 @@
+
+## Memory Analyser
+
+The memory analyser is a front for running and querying GDB about a state of
```
is a front for -> is a front-end used for

danpoe replied:
Probably we should only say here that the purpose of the memory analyzer is to take memory snapshots (of selects varibles), as it doesn't really provide features to make (custom) queries to GDB.

xbauch replied:
Changed.

danpoe commented on doc/cprover-manual/memory-analyzer.md:
```diff
@@ -0,0 +1,70 @@
+[CPROVER Manual TOC](../../)
+
+## Memory Analyser
```
Analyser -> Analyzer (and in other places in this PR)

xbauch replied:
Done.

sonodtt commented on doc/cprover-manual/memory-analyzer.md:
```diff
@@ -0,0 +1,70 @@
+We assume that the user wants to inspects a binary executable compiled with
+debugging symbols and a symbol-table information understandable by CBMC. That is
+most easily reproducible with (having `goto-gcc` on the `PATH`):
+
```
understandable by CBMC. [..]  -> understandable by CBMC, e.g. (having ...):
Succint, avoids the 'most easily reproducbile'

xbauch replied:
Changed.

sonodtt commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,70 @@
+$ goto-gcc -g -std=c11 input_program.c -o input_program_exe
+```
+
+The user can choose to either run the GDB from a core-file `--core-file cf` or
````
the GBD -> GDB

xbauch replied:
Done.

sonodtt commented on doc/cprover-manual/memory-analyzer.md:
```diff
@@ -0,0 +1,70 @@
+
+Finally, to obtain an output in JSON format, e.g. for further analyses by
+`goto-harness` pass the additional option `--json-ui` together with requesting
+the output to be a symbol table by `--symtab-snapshot`.
```
Please add an example command so its 100% clear. 

It might seem redunandant/excessive  - so here's my rationale: I've seen lots of documentation where people say 'just add this' and.. it does not work. You dig and finally they say: " oh I meant do this other thing and then add this".
So I prefer people to be quite specific. No possible misunderstandings.

xbauch replied:
No problem. Added.

danpoe commented on doc/cprover-manual/memory-analyzer.md:
```diff
@@ -0,0 +1,70 @@
+
+## Usage
+
+We assume that the user wants to inspects a binary executable compiled with
```
inspects -> inspect

xbauch replied:
Done.

danpoe commented on doc/cprover-manual/memory-analyzer.md:
```diff
@@ -0,0 +1,70 @@
+## Usage
+
+We assume that the user wants to inspects a binary executable compiled with
+debugging symbols and a symbol-table information understandable by CBMC. That is
```
a symbol-table -> symbol table

xbauch replied:
Done.

danpoe commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,70 @@
+$ goto-gcc -g -std=c11 input_program.c -o input_program_exe
+```
+
+The user can choose to either run the GDB from a core-file `--core-file cf` or
````
Can we say here that `goto-gcc` produces an ELF binary with a goto section that contains the goto model (goto program + symbol table)?

xbauch replied:
Sure.

danpoe commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,70 @@
+}
+```
+
+Finally, to obtain an output in JSON format, e.g. for further analyses by
````
analyses -> analysis

xbauch replied:
May be more than one.

danpoe commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -36,64 +39,141 @@ class gdb_apit;
+  ///   \ref get_expr_value on index expressions into \p expr.
+  /// \param expr: the expression to be analysed
+  /// \param array: array with zero-initialised elements
+  /// \param location: not used
```
I think we should here just say "the source location". The location is not directly used but it's nevertheless passed on to callees. (and same for other cases in this PR)

xbauch replied:
Done.

danpoe commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -36,64 +39,141 @@ class gdb_apit;
+  ///   expression associated with it.
+  /// Otherwise we add the appropriate \ref values mapping:
+  /// 1) call \ref gdb_apit::get_memory on the \p expr
+  ///    ??? Doesn't this have to be the same as \p memory_location
```
`get_memory()` in fact returns a struct with four members, one of them being the memory address. In this case we need the field `string` which is an `optionalt` that contains the string the pointer points to if it points to a string.

xbauch replied:
ok, thanks for the explanation.

Commit:
Add documentation to memory anayzer

Doxygen code-level description of main methods.

Commit:
Add entry to the manual wrt to memory analyzer

User-level covering running the analyzer and producing snapshots for
goto-harness.

hannes-steffenhagen-diffblue committed:
Add CMake function to set default CProver properties

We were doing this in multiple places, this should make it a bit easier to keep
set properties consistent among different places.

Ideally we could eventually move away from mentioning targets from modules here
on the top level (and instead have each module call cprover_default_properties)
but this isn't done in this commit.

hannes-steffenhagen-diffblue committed:
Make memory analyzer optional

Because memory analyzer depends on GDB being present and further uses platform
specific functionality at the moment it had some ifdef functionality to disable
itself.

This made the code a bit more complicated than it needed to be, and also lead to
the code effectively building defunct executables. This removes these ifdefs and
instead excludes memory-analyzer (and related tests) from the build unless
requested (via WITH_MEMORY_ANALYZER environment variable or CMake option
depending on whether it is a Make or CMake build respectively).

chrisr-diffblue reviewed (approved):
Looks like this addresses most of the comments - but its hard to track across PRs :-) Another pair of eyes would be good.

Commit:
Cumulative fix commit addressing comments

Adds Cmake regression tests for memory-analyzer, the gdb_api is not owned by the
analyzer, memory address is not a string, and other smaller changes.

Cleans up doxygen-comments in gdb_api and removes unused function
declaration (terminate_gdb_process) from gdb_api.h.

Renames the types_checking helpers in c_types_util.h

Fixes the counting of unit tests for Make.
